### DISCUSSION TITLE
How to deal special characters in a password?

### DISCUSSION THREAD

**Discussion by INailedIt** [SteamRE/DepotDownloader#186] [Q&A]
Is there a way to make depotdownloader read the steam login details from a separate file?

**Comment by psychonic** [comment]
This isn't really specific to DepotDownloader, and is entirely dependent on what shell you are using.

**Reply by INailedIt** [reply]
The game we're trying to download is only on windows so cmd I guess.

**Reply by psychonic** [reply]
Even on Windows, it can differ whether you're actually using cmd/bat, or PowerShell. Have you tried just putting the password in "double quotes"?

**Comment by FatalBulletHit** [comment]
_Disclaimer: This is targeted towards Windows OS._

# Summary
When passing a password with special characters as parameter it needs to be escaped on two levels.

Alternatively, you can omit the password parameter and DepotDownloader will prompt you to type (or paste) it without having to worry about escaping.

## TL;DR
**CMD**
Use this tool: https://output.jsbin.com/anitaz/11

**PowerShell**
Use this tool: https://output.jsbin.com/tirurof

# Details
## First level
The first level is the shell level. I will cover CMD and PowerShell - if you are using a different shell, different rules will apply.

### CMD
If you are using the CMD shell or a batch file (`.bat` or `.cmd`) the following special characters need to be escaped with a caret ( `^` ):

| Character to be escaped  | Escaped |
| - | - |
| ^ | ^^ |
| & | ^& |
| < | ^< |
| > | ^> |
| \| | ^\| |
| % | ^% |

**Examples**
| Password  | Escaped |
| - | -|
| somePassword | somePassword |
| some^Pass^^word | some^^Pass^^^^word |
| so&me^'Pass<wo%rd | so^&me^^'Pass^<wo^%rd |

**References**
You can read more about this on [SS64](https://ss64.com/nt/syntax-esc.html#escape).

### PowerShell
If you are using PowerShell or a PowerShell script (`.ps1`) you need to enclose the password with double ( `"` ) or single ( `'` ) quotes so PowerShell interprets it as a string.

For simplicity sake I recommended using single quotes as you will only need to replace each single quote within the password with an additional single quote.

**Examples**
| Password  | Escaped |
| - | -|
| somePassword | 'somePassword' |
| some'Password | 'some''Password' |
| so&me^'Pass<wo%rd | so&me^''Pass<wo%rd |

**References**
You can read more about this on [Microsoft Docs](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-7.1).

## Second level
The second level is the command line level. The command line is used to pass parameters to an application.

There are basically 3 rules:
- If a password contains spaces, it needs to be enclosed with double quotes (you may enclose it with double quotes)
- If a password contains double quotes, they need to be escaped by a preceding backslash ( `\` ).
- If a password contains backslashes, they don't need to be escaped unless they are preceding a double quote.

**Examples**
| Password  | Escaped |
| - | -|
| reallySafe password | "reallySafe password" |
| reallySafe"password | reallySafe\"password |
| reallySafe\password | reallySafe\password |
| reallySafe\"password | reallySafe\\\"password |
| really\"Safe\password electric"boogaloo\\" |"really\\\"Safe\password electric\"boogaloo\\\\"" |

Instead of doing this manually, you could use PowerShell and the .NET method [`String.Replace()`](https://docs.microsoft.com/en-us/dotnet/api/system.string.replace?view=net-5.0) to achieve this:

```
<# Password:
really\"Safe\password electric"boogaloo\\
#>

# Escape using single quotes
$password = 'really\"Safe\password electric"boogaloo\\"'

# Replace using .NET method String.Replace()
$password = $password.Replace('\"', '\\"').Replace('"', '\"')

.\DepotDownloader.exe -app 107410 -username Gaben -password $password
```

**References**
You can read more about this [here](https://daviddeley.com/autohotkey/parameters/parameters.htm) and [here](http://www.windowsinspired.com/understanding-the-command-line-string-and-arguments-received-by-a-windows-program/).
